Modified function for easy creation of banded matrices

GYeyosi opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->


#### References to other Issues or PRs
Fixes #16178 
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

In the above mentioned way of passing a dictionary giving diagonal number and the value,
I was successful in implementing it for dictionary values as just numbers or as lists in case if the values for the diagonal are not all the same. 
```
>>> Matrix.diag({0: 1, 1: 2},rows=3,cols=4)
Matrix([
[1, 2, 0, 0],
[0, 1, 2, 0],
[0, 0, 1, 2]])
>>> Matrix.diag({1: [1,2,3],0:1,-1:[-2,3]}, rows=3, cols=4)
Matrix([
[ 1, 1, 0, 0],
[-2, 1, 2, 0],
[ 0, 3, 1, 3]])
>>> Matrix.diag({1: [1,2,3],0:1,-1:[-2,3,4]}, rows=3, cols=4)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/matrices/common.py", line 773, in diag
    raise ValueError("The size of list provided should match the diagonal size")
ValueError: The size of list provided should match the diagonal size
```
when the value of dictionary is a function 
```
>>> Matrix.diag({1: [1,2,3],0:(lambda i, j: i + 1) ,-1:[-2,3]}, rows=3, cols=4)
Matrix([
[ 1, 1, 0, 0],
[-2, 2, 2, 0],
[ 0, 3, 3, 3]])
>>> Matrix.diag({-1: [-1, -2], 0: lambda i, j: i + j, 1: lambda d: 2*d + 1,2:4}, rows=3, cols=4)
Matrix([
[ 0,  1, 4, 0],
[-1,  2, 3, 4],
[ 0, -2, 4, 5]])
```

#### Other comments

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
 
* matrices
    * Added cases for checking the type of parameters passed and create matrix accordingly in the 
     Matrix.diag function of common.py 

    * Added the functionality where bands off the diagonal can be made by using a dictionary whose
        keys tell the diagonal

<!-- END RELEASE NOTES -->

GYeyosi committed:
Modified function for easy creation of banded matrices

smichr commented on sympy/matrices/common.py:
```diff
@@ -732,6 +732,8 @@ def size(m):
             """Compute the size of the diagonal block"""
             if hasattr(m, 'rows'):
                 return m.rows, m.cols
+            if isinstance(m,dict):
```
Looking good.

Please do PEP8 tidy (e.g. spaces after commas, space around `+-` operators, comparisons like ` == `, and after `if`)

smichr commented on sympy/matrices/common.py:
```diff
@@ -753,6 +755,40 @@ def size(m):
                         diag_entries[(i + row_pos, j + col_pos)] = m[i, j]
                 row_pos += m.rows
                 col_pos += m.cols
+            elif(isinstance(m,dict)):
```
outer parens not necessary

smichr commented on sympy/matrices/common.py:
```diff
@@ -753,6 +755,40 @@ def size(m):
+            elif(isinstance(m,dict)):
+                # in this case we're a dict
+                for key,value in m.items():
+                    r_p = 0
```
since you redefine these in the next two lines, these two lines are not necessary

smichr commented on sympy/matrices/common.py:
```diff
@@ -753,6 +755,40 @@ def size(m):
+                # in this case we're a dict
+                for key,value in m.items():
+                    r_p = 0
+                    c_p = 0
```
key should be an int, so `key = as_int(key)` will check for that

smichr commented on sympy/matrices/common.py:
```diff
@@ -753,6 +755,40 @@ def size(m):
+                for key,value in m.items():
+                    r_p = 0
+                    c_p = 0
+                    r_p = 0 if key>0 else -(key)
```
-(key) -> -key

smichr commented on sympy/matrices/common.py:
```diff
@@ -753,6 +755,40 @@ def size(m):
+                                r_p += 1
+                                c_p += 1
+                        elif(isfunction(value)):
+                            num_args = len(getargspec(value).args)
```
import _getnargs from sympy.core.function, then: `num_args = _getnargs(value)`

GYeyosi replied:
I'm having some problem in using it that way. It returned None. Can you please help me out.

When I use python version 2.7
```
Python 2.7.15+ (default, Feb  3 2019, 13:13:16) 
[GCC 8.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.core.function import _getnargs
>>> def func(a,b,c):
...     return a+b+c
... 
>>> a = _getnargs(func)
>>> print(a)
None
>>> a = _getnargs(lambda i, j: i + 1)
>>> print(a)
None
>>> 
>>> 
>>> 
>>> from inspect import isfunction, getargspec
>>> a = num_args = len(getargspec(func).args)
>>> print(a)
3
>>> a = num_args = len(getargspec(lambda i, j: i + 1).args)
>>> print(a)
2
>>> 
```
Python version 3.7
```
Python 3.7.2+ (default, Feb  2 2019, 14:31:48) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.core.function import _getnargs
>>> def func(a,b,c):
...     return a+b+c
... 
>>> a = _getnargs(func)
>>> print(a)
None
>>> a = _getnargs(lambda i, j: i + 1)
>>> print(a)
None
>>> 
>>> 
>>> 
>>> from inspect import signature
>>> a = len(signature(func).parameters)
>>> print(a)
3
>>> a = len(signature(lambda i, j: i + 1).parameters)
>>> print(a)
2
>>> 
>>> 

```

smichr replied:
> Can you please help me out.

```diff
diff --git a/sympy/core/function.py b/sympy/core/function.py
index 74fd41c..23aa952 100644
--- a/sympy/core/function.py
+++ b/sympy/core/function.py
@@ -109,6 +109,11 @@ def _getnargs(cls):
             return _getnargs_old(cls.eval)
         else:
             return _getnargs_new(cls.eval)
+    elif inspect.isfunction(cls):
+        if sys.version_info < (3, ):
+            return len(inspect.getargspec(cls).args)
+        else:
+            return len(inspect.signature(cls).parameters)
     else:
         return None

diff --git a/sympy/core/tests/test_function.py b/sympy/core/tests/test_function.py
index 130e19b..7937643 100644
--- a/sympy/core/tests/test_function.py
+++ b/sympy/core/tests/test_function.py
@@ -5,7 +5,7 @@
         Matrix, Basic)
 from sympy.utilities.pytest import XFAIL, raises
 from sympy.core.basic import _aresame
-from sympy.core.function import PoleError, _mexpand
+from sympy.core.function import PoleError, _mexpand, _getnargs
 from sympy.core.sympify import sympify
 from sympy.sets.sets import FiniteSet
 from sympy.solvers.solveset import solveset
@@ -18,6 +18,7 @@
 f, g, h = symbols('f g h', cls=Function)
 _xi_1, _xi_2, _xi_3 = [Dummy() for i in range(3)]

+
 def test_f_expand_complex():
     x = Symbol('x', real=True)

@@ -158,6 +159,15 @@ def test_nargs():
     raises(ValueError, lambda: Function('f', nargs=()))


+def test__getnargs():
+    f = lambda x, y: 1
+    assert _getnargs(f) == 2
+    def f(x, y):
+        pass
+    assert _getnargs(f) == 2
+    assert _getnargs(log) == (1, 2)
+
+
 def test_Lambda():
     e = Lambda(x, x**2)
     assert e(4) == 16
```

GYeyosi replied:
Thank you, I've made the changes.
Can you please review it now?

smichr commented on sympy/matrices/common.py:
```diff
@@ -753,6 +755,40 @@ def size(m):
+                                diag_entries[(r_p,c_p)] = value(r_p,c_p)
+                                r_p+=1
+                                c_p+=1
+                            if(num_args==1):
```
if->elif and then the two increments (e.g. r_p += 1) can be moved to after the if-elif block

GYeyosi committed:
Edits made as suggested

GYeyosi committed:
Edit getnargs and added tests

smichr commented:
Now we have to make sure that it plays nicely with other diagonal elements:

```
>>> M=Matrix;M.diag(11,22,{1:1,3:3,-3:-2,'size':(5,5)},33)
Matrix([
[11,  0,  0,  0, 0, 0, 0,  0],
[ 0, 22,  0,  0, 0, 0, 0,  0],
[ 0,  0,  0,  1, 0, 3, 0,  0],
[ 0,  0,  0,  0, 1, 0, 3,  0],
[ 0,  0,  0,  0, 0, 1, 0,  0],
[ 0,  0, -2,  0, 0, 0, 0,  0],
[ 0,  0,  0, -2, 0, 0, 0,  0],
[ 0,  0,  0,  0, 0, 0, 0, 33]])
```
Something like this, maybe?
```diff
diff --git a/sympy/matrices/common.py b/sympy/matrices/common.py
index 94589d6..2428b18 100644
--- a/sympy/matrices/common.py
+++ b/sympy/matrices/common.py
@@ -7,6 +7,7 @@
 from __future__ import division, print_function
 
 from collections import defaultdict
+from inspect import isfunction
 from types import FunctionType
 
 from sympy.assumptions.refine import refine
@@ -15,13 +16,14 @@
     Iterable, as_int, is_sequence, range, reduce)
 from sympy.core.decorators import call_highest_priority
 from sympy.core.expr import Expr
-from sympy.core.function import count_ops
+from sympy.core.function import count_ops, _getnargs
 from sympy.core.singleton import S
 from sympy.core.symbol import Symbol
 from sympy.core.sympify import sympify
 from sympy.functions import Abs
 from sympy.simplify import simplify as _simplify
 from sympy.utilities.iterables import flatten
+from sympy.utilities.misc import filldedent
 
 
 class MatrixError(Exception):
@@ -733,11 +735,21 @@ def size(m):
             if hasattr(m, 'rows'):
                 return m.rows, m.cols
             if isinstance(m, dict):
-                return kwargs.get('rows'), kwargs.get('cols')
+                rv = m.pop('size', (kwargs.get('rows'), kwargs.get('cols')))
+                if None in rv:
+                    raise ValueError(filldedent('''
+                        If size is not given in dict the `rows` and `cols`
+                        must be given as arg to diag.'''))
+                return rv
             return 1, 1
-
-        diag_rows = sum(size(m)[0] for m in args)
-        diag_cols = sum(size(m)[1] for m in args)
+        dict_sizes = []
+        diag_rows = diag_cols = 0
+        for m in args:
+            s = r, c = size(m)
+            if isinstance(m, dict):
+                dict_sizes.append(s)
+            diag_rows += r
+            diag_cols += c
         rows = kwargs.get('rows', diag_rows)
         cols = kwargs.get('cols', diag_cols)
         if rows < diag_rows or cols < diag_cols:
@@ -748,55 +760,62 @@ def size(m):
         # fill a default dict with the diagonal entries
         diag_entries = defaultdict(lambda: S.Zero)
         row_pos, col_pos = 0, 0
+        ndict = 0
         for m in args:
-            if hasattr(m, 'rows'):
-                # in this case, we're a matrix
+            if hasattr(m, 'rows'):  # a matrix
                 for i in range(m.rows):
                     for j in range(m.cols):
                         diag_entries[(i + row_pos, j + col_pos)] = m[i, j]
                 row_pos += m.rows
                 col_pos += m.cols
-            elif isinstance(m, dict):
-                # in this case we're a dict
+            elif not isinstance(m, dict):  # a single value
+                diag_entries[(row_pos, col_pos)] = m
+                row_pos += 1
+                col_pos += 1
+            else:  # a dict
+                ix = lambda r, c: (row_pos + r, col_pos + c)
+                r, c = dict_sizes[ndict]
+                rmax = r - 1
+                cmax = c - 1
                 for key, value in m.items():
                     key = as_int(key)
-                    assert(key < diag_rows and key < diag_cols)
                     r_p = 0 if key >= 0 else -key
                     c_p = key if key > 0 else 0
-                    func_arg = 0
-                    while (r_p < diag_rows and c_p < diag_cols):
-                        from inspect import isfunction
-                        from sympy.core.function import _getnargs
-
+                    if r_p > rmax or c_p > cmax:
+                        raise ValueError(filldedent('''
+                            dict-specified diagonal index out of range'''))
+                    d = 0
+                    dlen = d + 1
+                    while (r_p < diag_rows and c_p < diag_cols and d < dlen):
+                        dlen = min(rmax - r_p + 1, cmax - c_p + 1)
                         if(isinstance(value, list)):
-                            if(len(value) != min(diag_rows - r_p, diag_cols - c_p)):
-                                raise ValueError("The size of list provided should match the diagonal size")
+                            if(len(value) != dlen):
+                                raise ValueError(filldedent('''
+                                    The size of list provided should
+                                    match the diagonal size.'''))
                             for i in range(len(value)):
-                                diag_entries[(r_p, c_p)] = value[i]
+                                diag_entries[ix(r_p, c_p)] = value[i]
                                 r_p += 1
                                 c_p += 1
                         elif(isfunction(value)):
                             num_args = _getnargs(value)
-                            # print(num_args)
-                            if(num_args == 2):
-                                diag_entries[(r_p, c_p)] = value(r_p, c_p)
-                            elif(num_args == 1):
-                                diag_entries[(r_p, c_p)] = value(func_arg)
-                                func_arg += 1
+                            if num_args == 2:
+                                diag_entries[ix(r_p, c_p)] = value(r_p, c_p)
+                            elif num_args == 1:
+                                diag_entries[ix(r_p, c_p)] = value(d)
                             else:
-                                raise ValueError("While passing function as a value in dictionary number of arguments must be either 1 or 2")
-                            r_p += 1
-                            c_p += 1
+                                raise ValueError(filldedent('''
+                                    The functions in dict-described
+                                    diagonals must have 1 or 2 args.'''))
                         else:
-                            diag_entries[(r_p, c_p)] = value
-                            r_p += 1
-                            c_p += 1
-
-            else:
-                # in this case, we're a single value
-                diag_entries[(row_pos, col_pos)] = m
-                row_pos += 1
-                col_pos += 1
+                            diag_entries[ix(r_p, c_p)] = value
+                        r_p += 1
+                        c_p += 1
+                        d += 1
+                r, c = dict_sizes[ndict]
+                row_pos += r
+                col_pos += c
+                ndict += 1
         return klass._eval_diag(rows, cols, diag_entries)
 
     @classmethod
```

GYeyosi commented:
I didn't consider the case, where dictionary is constricted with some **size**.
This looks great, I'll add it.

GYeyosi committed:
Make sure it works fine if we add other diagonal elements other than dictionary

smichr commented:
Perhaps the dict could be a little flexible in size:
```diff
diff --git a/sympy/matrices/common.py b/sympy/matrices/common.py
index 2428b18..b2fb3ff 100644
--- a/sympy/matrices/common.py
+++ b/sympy/matrices/common.py
@@ -739,7 +739,7 @@ def size(m):
                 if None in rv:
                     raise ValueError(filldedent('''
                         If size is not given in dict the `rows` and `cols`
-                        must be given as arg to diag.'''))
+                        must be given as args to diag.'''))
                 return rv
             return 1, 1
         dict_sizes = []
@@ -753,9 +753,17 @@ def size(m):
         rows = kwargs.get('rows', diag_rows)
         cols = kwargs.get('cols', diag_cols)
         if rows < diag_rows or cols < diag_cols:
-            raise ValueError("A {} x {} diagnal matrix cannot accommodate a"
-                             "diagonal of size at least {} x {}.".format(rows, cols,
-                                                                         diag_rows, diag_cols))
+            if len(dict_sizes) == 1 and len(args) > 1:
+                r, c = dict_sizes[0]
+                dict_sizes = [
+                    (r - (diag_rows - rows), c - (diag_cols - cols))]
+                diag_rows = rows
+                diag_cols = cols
+            else:
+                raise ValueError(filldedent('''
+                    The diagonal elements need a matrix that is {} x {}
+                    but only {} x {} has been specified.'''.format(
+                    diag_rows, diag_cols, rows, cols)))
 
         # fill a default dict with the diagonal entries
         diag_entries = defaultdict(lambda: S.Zero)
```
Here are some possible tests:
```python
>>> Matrix.diag(1, {1:2}, rows=3, cols=4)
Matrix([
[1, 0, 0, 0],
[0, 0, 2, 0],
[0, 0, 0, 2]])
>>> Matrix.diag(1, {1: 2, 'size': (2, 2)}, rows=3, cols=4)
Matrix([
[1, 0, 0, 0],
[0, 0, 2, 0],
[0, 0, 0, 0]])
>>> Matrix.diag(1, {1: 2, 'size': (2, 2)}, 3, rows=4, cols=4)
Matrix([
[1, 0, 0, 0],
[0, 0, 2, 0],
[0, 0, 0, 0],
[0, 0, 0, 3]])
>>> Matrix.diag(1, {1:2}, 3, rows=5, cols=5)
Matrix([
[1, 0, 0, 0, 0],
[0, 0, 2, 0, 0],
[0, 0, 0, 2, 0],
[0, 0, 0, 0, 0],
[0, 0, 0, 0, 3]])
>>> Matrix.diag(1, {-1: 2}, 3, rows=5, cols=5)
Matrix([
[1, 0, 0, 0, 0],
[0, 0, 0, 0, 0],
[0, 2, 0, 0, 0],
[0, 0, 2, 0, 0],
[0, 0, 0, 0, 3]])
```

oscargus commented on sympy/core/function.py:
```diff
@@ -109,6 +109,11 @@ def _getnargs(cls):
         else:
             return _getnargs_new(cls.eval)
+    elif inspect.isfunction(cls):
+        if sys.version_info < (3, ):
```
You can import `PY3` from `sympy.core.compatibility` and use here (although the if should be reversed then...).

GYeyosi commented:
What should be the behaviour when user provides with an input of this type? 
```
>>> Matrix.diag({0:1}, {1:2}, rows=3, cols=4)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/matrices/common.py", line 767, in diag
    diag_rows, diag_cols, rows, cols)))
ValueError: 
The diagonal elements need a matrix that is 6 x 8 but only 3 x 4 has
been specified.
```
Should we merge dictionaries? and make it 
```
>>> Matrix.diag({0: 1, 1: 2},rows=3,cols=4)
Matrix([
[1, 2, 0, 0],
[0, 1, 2, 0],
[0, 0, 1, 2]])
>>> 

```

GYeyosi committed:
Edits done to handle flexible dict size

smichr commented:
> Should we merge dictionaries? and make it

Interesting that the un-specified size goes to the end of the diagonal whereas 
```python
>>> Matrix.diag({0: 1, 1: 2,'size':(3,3)},rows=3,cols=4)
Matrix([
[1, 2, 0, 0],
[0, 1, 2, 0],
[0, 0, 1, 0]])
```
This concatenation of contiguous dicts makes a nice method of being able to say what the diagonal should be, e.g.
```
>>> ones = {1:1, -1:1}
>>> Matrix.diag(ones, {0:2*x}, rows=4, cols=4)
Matrix([
[2*x,   1,   0, 0],
[  1, 2*x,   1, 0],
[  0,   1, 2*x, 1],
[  0,   0,   1, 0]])
```

GYeyosi commented:
> Interesting that the un-specified size goes to the end of the diagonal whereas
> 
> ```python
> >>> Matrix.diag({0: 1, 1: 2,'size':(3,3)},rows=3,cols=4)
> Matrix([
> [1, 2, 0, 0],
> [0, 1, 2, 0],
> [0, 0, 1, 0]])
> ```
What about this then, from the above discussions.
```
>>> Matrix.diag(1, {-1: 2}, 3, rows=5, cols=5)
Matrix([
[1, 0, 0, 0, 0],
[0, 0, 0, 0, 0],
[0, 2, 0, 0, 0],
[0, 0, 2, 0, 0],
[0, 0, 0, 0, 3]])
```
I think, I lost my clarity on what exactly a size key in dictionary says.
Can you please help me out.

smichr commented:
After collapsing contiguous dictionaries, if there is more than one dictionary without 'size' key then raise an error.

Something like this, perhaps?
```python
newargs = []
nosize = 0
for a in args:
    if newargs and all(isinstance(i, dict) and 'size' not in i for i in (a, newargs[-1])):
        newargs[-1].updated(a)
    else:
        newargs.append(a)
        if isinstance(a, dict) and 'size' not in a:
            nosize += 1
if nosize > 1:
    raise ValueError(filldedent('''
        non-contiguous dictionaries must have a size specified,
        e.g. {'size': (rows, cols), ...}'''))
```
I was just checking on [special block diagonal shapes](https://en.wikipedia.org/wiki/Block_matrix#Block_diagonal_matrices) and see that the "direct sum" is computed with this routine:
```python
>>> A=Matrix(((1,2,3),(4,5,6)))
>>> B=Matrix(((3,3),(4,4)))
>>> Matrix.diag(A, B)  # direct sum
Matrix([
[1, 2, 3, 0, 0],
[4, 5, 6, 0, 0],
[0, 0, 0, 3, 3],
[0, 0, 0, 4, 4]])
>>>
```

GYeyosi committed:
Functionality for non-contiguous dictionaries

smichr commented:
If there is only one dictionary and neither size, rows nor cols, the minimal size can be used:
```python
def minsize(d):
    r = max(0, -min(d)) + 1
    c = -min(0, -max(d)) + 1
    for k in d:
        size = len(d[k]) if type(d[k]) is list else 0
        r = max(r, abs(k) + size)
        c = max(c, abs(k) + size)
    return r, c

>>> minsize({2:[4,5,6],-2:[1,2,3]})
(5, 5)
>>> Matrix.diag({1:2,-2:1})
Matrix([
[0, 2],
[0, 0],
[1, 0]])
```

GYeyosi commented:
> If there is only one dictionary and neither size, rows nor cols, the minimal size can be used:
 
Yup, I'm actually implementing this now. I'll add it and push the changes.

GYeyosi committed:
Case of only one dictionary and neither size, rows nor cols provided

GYeyosi commented:
> Yup, I'm actually implementing this now. I'll add it and push the changes.

Done with it, pushed the changes.
@smichr can you please review it.

smichr commented:
I noticed that 1s were trailing out of range in what should be
```
>>> Matrix.diag({1:2,-2:1},4)
Matrix([
[0, 2, 0],
[0, 0, 0],
[1, 0, 0],
[0, 0, 4]])
```
I modified that and tweaked what you have, too.

smichr committed:
watch for assigment past dict's size

Also, fill in sizes before computing diagonal size and raise errors as needed.

GYeyosi commented on sympy/matrices/common.py:
```diff
@@ -726,78 +726,99 @@ def diag(kls, *args, **kwargs):
+                e.g. {'size': (rows, cols), ...}.'''))
+        elif nosize:  # length of 1
+            i = nosize[0]
+            if None is (rows, cols):
```
Shoudn't it be (if None in) instead of (if None is).
`if None in (rows, cols):`

GYeyosi commented on sympy/matrices/common.py:
```diff
@@ -726,78 +726,99 @@ def diag(kls, *args, **kwargs):
+                    spec_cols += c
+                rneed = rows - spec_rows
+                cneed = cols - spec_cols
+                r, c = dsize
```
dsize is not assigned any value. It show error in following case.

`Matrix.diag({1:2,-2:1},4, rows = 3, cols = 4)`
I've added assignment of it.

GYeyosi committed:
Added missing assignment of dsize.

GYeyosi commented:
> I was just checking on [special block diagonal shapes](https://en.wikipedia.org/wiki/Block_matrix#Block_diagonal_matrices) and see that the "direct sum" is computed with this routine:

The DirectSum can be done as below.
```
class DirectSum(MatrixExpr):

    def __new__(cls, *args):
        if not args:
            return GenericZeroMatrix()

        # This must be removed aggressively in the constructor to avoid
        # TypeErrors from GenericZeroMatrix().shape
        args = filter(lambda i: GenericZeroMatrix() != i, args)
        args = list(map(sympify, args))
        validate(*args)
        obj = Basic.__new__(cls, *args)
        return obj

    def doit(self):
        return Matrix.diag(self.args)

def validate(*args):
    if not all(arg.is_Matrix for arg in args):
        raise TypeError("Mix of Matrix and Scalar symbols")
```
Should I add it to expressions ?

smichr committed:
remove 'size' before using min/max

smichr committed:
don't use i in compreh'n and as counter

smichr committed:
doctest update and tests

smichr committed:
func_name strips off path

smichr committed:
use func_name in common docstring

smichr committed:
dict entries cannot be Matrix

smichr commented:
> Should I add it to expressions ?

I defer to others more familiar with Matrices.

smichr commented:
Rather than modify this directly here,  I made a [PR to this branch](https://github.com/GYeyosi/sympy/pull/1) which I think you will be able to merge (or comment on before doing so).

sylee957 commented:
The function had grown too large. Is it possible to split some parts?

smichr committed:
reorganize

smichr committed:
Update common.py

smichr committed:
Update test_commonmatrix.py

smichr committed:
Update common.py

GYeyosi committed:
Merge pull request #1 from smichr/diag

smichr commented on sympy/matrices/common.py:
```diff
@@ -753,11 +884,39 @@ def size(m):
+                            for i in range(len(value)):
+                                diag_entries[D] = value[i]
+                                D = D[0] + 1, D[1] + 1
+                            r_p = rmax
```
these two lines can just be `break` to get out of while loop

GYeyosi replied:
Yes, that can be done, I edited it to `break`.

GYeyosi committed:
update common.py

asmeurer commented:
The release notes entry should mention which function is actually affected by the change (Matrix.diag).

GYeyosi commented:
> The release notes entry should mention which function is actually affected by the change (Matrix.diag)

I've added it to the Release Notes.

smichr commented:
The only other thing I was thinking about was a more flexible way to position the filling. Instead of using kerning matrices (which only allow nonnegative motion) we could use keywords 'goto' and 'move' --or something like that--to indicate absolute position or relative motion, e.g. `diag(dict(move=(-2,0)), 1, 2, 3,dict(goto=(0,2)),4,5,6)`

GYeyosi commented:
> The only other thing I was thinking about was a more flexible way to position the filling. Instead of using kerning matrices (which only allow nonnegative motion) we could use keywords 'goto' and 'move' --or something like that--to indicate absolute position or relative motion, e.g. `diag(dict(move=(-2,0)), 1, 2, 3,dict(goto=(0,2)),4,5,6)`

I'm not sure if I understood what you meant exactly, is it like not using the integral keys in the dictionary but use these keywords to specify the position to start the filling? 
Can you please give an example explaining this.

smichr commented:
This `diag(dict(move=(-2,0)), 1, 2, 3,dict(goto=(0,2)),4,5,6)` would recreate the docstring example in which the kerning is used.

GYeyosi commented:
Consider this, `diag({1:2, -2:1, 'move':(-2, 0)}, 1, 2, 3,dict(goto=(0, 2)), 4,5,6)`
In the first dictionary, should I move the values down (2 rows) only of that dictionary, or the values 1, 2, 3 also must be moved? and in similar case with goto also.

smichr commented:
the motion directives would be singletons; they would not appear inside another dictionary. You can see my wip on my diag branch. I haven't yet settled how to compute size so the example I have been giving fails.

GYeyosi commented:
I have seen the branch you mentioned, in your code the arg_size function can be written this way.
I was working on code from your branch, successfully implemented it for arguments as  values, i'm now extending it for matrices, dicts. Also, there are many unnecessary lines in it, needed  to be modified.
```
def arg_size(args):                    
    prev = None
    o_r, o_c, g_r, g_c, m_r, m_c = 0, 0, 0, 0, 0, 0
    max_row, max_col = 0, 0
    for a_ in args:
        a_size = size(a_)
        if type(a_) is dict and 'goto' in a_:
            prev = 'goto'
            prev_size = a_size
            g_r , g_c = prev_size[0], prev_size[1]
            continue
        elif type(a_) is dict and 'move' in a_:
            prev = 'move'
            prev_size = a_size
            continue
        else: 
            o_r += a_size[0]
            o_c += a_size[1]
            if prev is None:
                continue
            if 'goto' in prev:
                g_r += a_size[0]
                g_c += a_size[1]
                if max_row < g_r:
                    max_row = g_r 
                if max_col < g_c:
                    max_col = g_c 
            elif 'move' in prev:
                m_r = o_r - prev_size[0]                    
                m_c = o_c - prev_size[1]
                # print(m_r, m_c)
                if max_row < m_r:
                    max_row = m_r 
                if max_col < m_c:
                    max_col = m_c 
    return max_row, max_col
```

GYeyosi committed:
Functionality for more flexible way to position the filling

GYeyosi commented:
@smichr Can you please review the changes I've just made. 
Although the code is lengthy now, I hope all possible cases are covered.
Can you please check if any test case fails.

GYeyosi committed:
strip_white space characters

smichr commented:
Does this do the same?
```python
        def arg_size(args):
            prev = None
            max_row, max_col, o_r, o_c = [0]*4
            for a_ in args:
                a_size = size(a_)
                if type(a_) is dict and 'goto' in a_ or 'move' in a_:
                    if 'goto' in a_:
                        prev = 'goto'
                    else:
                        prev = 'move' 
                    prev_size = a_size
                else:
                    R, C = a_size
                    o_r += R
                    o_c += C
                    if prev is None:
                        max_row = o_r
                        max_col = o_c
                    else:
                        r, c = prev_size
                        if prev == 'goto':
                            max_row = max(max_row, R + r)
                            max_col = max(max_col, C + c)
                        else:  # 'move'
                            max_row = max(max_row, o_r - r)
                            max_col = max(max_col, o_c - c)
            return max_row, max_col
```

GYeyosi commented:
> Does this do the same?

I changed the line `if type(a_) is dict and 'goto' in a_ or 'move' in a_:`  to 
`if type(a_) is dict and ('goto' in a_ or 'move' in a_):` cause the first one raises error if a_ is an integer.

It gave the following result for `diag(dict(move=(-2,0)), 1, 2, 3,dict(goto=(0,2)),4,5,6)` of size `(5,3)` whereas it should be `(5,5)`
```
>>> diag(dict(move=(-2,0)), 1, 2, 3,dict(goto=(0,2)),4,5,6)
Matrix([
[0, 0, 4],
[0, 0, 0],
[1, 0, 0],
[0, 2, 0],
[0, 0, 3]])
```

GYeyosi committed:
update common.py

GYeyosi commented:
I've tried simplifying the code
Can you have a look at it and suggest further modifications that can be done.

smichr commented:
```python
        def arg_size(args):
            prev = None
            max_row, max_col, o_r, o_c, g_r, g_c = [0]*6
            for a_ in args:
                prev_size = a_size = size(a_)
                if type(a_) is dict and 'goto' in a_:
                    prev = 'goto'
                    g_r , g_c = a_size
                elif type(a_) is dict and 'move' in a_:
                    prev = 'move'
                else:
                    o_r += a_size[0]
                    o_c += a_size[1]
                    if prev is None:
                        max_row, max_col = o_r, o_c
                    elif 'goto' in prev:
                        g_r += a_size[0]
                        g_c += a_size[1]
                        max_row = max(max_row, g_r)
                        max_col = max(max_col, g_c)
                    elif 'move' in prev:
                        max_row = max(max_row, o_r - prev_size[0])
                        max_col = max(max_col, o_c - prev_size[1])
            return max_row, max_col
```

GYeyosi commented:
> ```python
>prev_size = a_size = size(a_)
>if type(a_) is dict and 'goto' in a_:
>    prev = 'goto'
>    g_r , g_c = a_size
> elif type(a_) is dict and 'move' in a_:
>     prev = 'move'
> ```

prev_size should be changed only if a_ is `move`or `goto`, so I've made it
```
a_size = size(a_)
if type(a_) is dict and 'goto' in a_:
    prev = 'goto'
    prev_size = size(a_)
    g_r , g_c = a_size
elif type(a_) is dict and 'move' in a_:
    prev = 'move'
    prev_size = size(a_)
```

smichr commented:
> prev_size should be changed only

just saw that -- see update code that I last posted

GYeyosi commented:
I've pushed the changes.

GYeyosi committed:
update common.py

smichr commented on sympy/matrices/common.py:
```diff
@@ -732,32 +756,226 @@ def size(m):
 
         # fill a default dict with the diagonal entries
         diag_entries = defaultdict(lambda: S.Zero)
         row_pos, col_pos = 0, 0
```
This looks like repetition of `arg_size`

smichr commented on sympy/matrices/common.py:
```diff
@@ -732,32 +756,226 @@ def size(m):
+            prev = None
+            max_row, max_col, o_r, o_c, g_r, g_c = [0]*6
+            for a_ in args:
+                a_size = size(a_)
```
```
k = kern(a)
a_size = size(a_) if k is None else is k
if k:
    prev_size = a_size
...
```

smichr commented on sympy/matrices/common.py:
```diff
@@ -732,32 +756,226 @@ def size(m):
+                            or Expr in dict'''))
+
+        def kern(d):
+            if 'move' in d or 'goto' in d:
```
--> ```if type(d) is dict and ('move' in d or 'goto' in d):```

GYeyosi replied:
Not required.. I guess cause we use `kern(d)` only when d is a dict.

GYeyosi committed:
update common.py

GYeyosi committed:
update common.py

smichr committed:
use classof: not all obj have shape

smichr committed:
recognize kerning dicts

smichr committed:
add tests; clean up test_commonmatrix

GYeyosi committed:
Merge pull request #2 from smichr/diag

Diag

GYeyosi commented:
Looks good.

smichr committed:
Merge branch 'master' into master

smichr committed:
use isfunction

smichr committed:
replace assertions with ValueErrors

smichr committed:
16233: modify _MinimalMatrix.__init__

smichr commented:
I got the issue #16233 fixed and added ValueErrors instead of Assertions. coverage of diag was 100% in terms of line coverage the last time I checked (but for some reason the report-writing is not working for the coverage test and I can't check it).

GYeyosi committed:
Merge pull request #3 from smichr/diag

smichr commented on sympy/matrices/common.py:
```diff
@@ -701,50 +705,218 @@ def diag(kls, *args, **kwargs):
+                if len(d) != 1:
+                    raise ValueError(filldedent('''
+                        A kerning dict should not contain any other
+                        items; do not inlude kerning directives in
```
when the next changes are made, 'inlude' -> 'include'

smichr commented:
Let's see if anyone else has some ideas about this approach.

smichr commented:
I see there is still a problem.
```python
>>> diag({-2:1}, row=4,cols=4)
Matrix([
[0, 0, 0, 0],
[0, 0, 0, 0],
[1, 0, 0, 0]])

should be

Matrix([
[0, 0, 0, 0],
[0, 0, 0, 0],
[1, 0, 0, 0],
[0, 1, 0, 0]])
```

GYeyosi commented:
> I see there is still a problem.
> ```python
> >>> diag({-2:1}, row=4,cols=4)
> Matrix([
> [0, 0, 0, 0],
> [0, 0, 0, 0],
> [1, 0, 0, 0]])
> 
> ```
The keyword to define number of rows is `rows` not `row`

oscarbenjamin commented:
This API seems overly complex. It has taken me some time to understand it right now.

As an example:
```
        A kerning dictionary can be used to reposition the filling:
        {'move': (row change, col change)},
        {'goto': (row, col)}, or
        {} (to go to the diagonal):

        >>> diag(7, {'move': (-1, -1)}, 1, 2, 3, {'goto': (0, 2)}, 4, 5, 6, {}, 8)
        Matrix([
        [7, 0, 4, 0, 0, 0],
        [0, 0, 0, 5, 0, 0],
        [1, 0, 0, 0, 6, 0],
        [0, 2, 0, 0, 0, 0],
        [0, 0, 3, 0, 0, 0],
        [0, 0, 0, 0, 0, 8]])
```
I can't really imagine wanting to do something like this and I wouldn't want to try and understand code that was written this way. Wouldn't this example be better written as this:
```
diag({0: [7]+4*[0], -2:[1,2,3], 2:[4,5,6]}, 8)
```

Can we draw inspiration from other APIs here like numpy/scipy/Matlab? I don't think any of those has such a state machine API: rather these effects are achieved through composition of many simple primitives.

Is it intentional that I can do this:
```julia
In [5]: Matrix.diag(1, 2, {'move':(1,-1)}, {0:3})                                                                                                             
Out[5]: 
⎡1  0⎤
⎢    ⎥
⎣0  3⎦
```
It took me a couple of goes to get the indices right for that move. Is the coordinate system +/+ for upwards and to the right? That makes the horizontal direction alined with increasing indices and the vertical direction reversed.

GYeyosi commented:
> I can't really imagine wanting to do something like this and I wouldn't want to try and understand code that was written this way. Wouldn't this example be better written as this:
> ```
> diag({0: [7]+4*[0], -2:[1,2,3], 2:[4,5,6]}, 8)
> ``` 
Either of the two gives out the same matrix as result, but in the former one we use keywords 'goto' and 'move' to indicate absolute position or relative motion whereas latter uses kerning matrices (allow only non-negative motion).

> Is it intentional that I can do this:
> ```julia
> In [5]: Matrix.diag(1, 2, {'move':(1,-1)}, {0:3})                                                                                                             
> Out[5]: 
> ⎡1  0⎤
> ⎢    ⎥
> ⎣0  3⎦
> ```
Yes, I think it should be this way cause it indicates the relative motion of the block `{0: 3}` to a step left and a step up. 
> Is the coordinate system +/+ for upwards and to the right? That makes the horizontal direction alined with increasing indices and the vertical direction reversed.

The coordinate system we chose as of now:
1. If the key is `goto` we need to start filling from than position (index).
2. If the key is `move` and value `(x, y)` we'll relatively move the block x rows up and y rows right.
This is the reason in order to move the block down we need to give a negative x. (This convention can be changed if required). 

@smichr  Hope I'm correct with the conventions mentioned above, can you please crosscheck.

smichr commented:
> can you please crosscheck.

yes, but I agree with @oscarbenjamin . I think we can get rid of the kerning.

GYeyosi committed:
Removed the functionality of kerning using keywords 'goto' and 'move'

GYeyosi commented:
I've reset the code back to the one in which we didn't implement any functionality for `move` and `goto`.

GYeyosi committed:
Update common.py

oscarbenjamin commented:
I find diag useful for creating block diagonal matrices:
```
In [1]: diag(ones(2), ones(2))                                                                                                                                
Out[1]: 
⎡1  1  0  0⎤
⎢          ⎥
⎢1  1  0  0⎥
⎢          ⎥
⎢0  0  1  1⎥
⎢          ⎥
⎣0  0  1  1⎦
```

Does this PR affect that? Also do you think there's a reasonable way of using the features here for banded block matrices?

I just tried this Pr with:
```
In [2]: diag({-1:ones(2)}, ones(2))                                                                                                                           
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
```

GYeyosi commented:
> I find diag useful for creating block diagonal matrices: 
> In [1]: diag(ones(2), ones(2))                                                                                                                    
> Does this PR affect that? 

Nope, even now diag is useful for creating block diagonal matrices.

> Also do you think there's a reasonable way of using the features here for banded block matrices?
> I just tried this Pr with:
> 
> ```
> In [2]: diag({-1:ones(2)}, ones(2))                                                                                                                           
> ---------------------------------------------------------------------------
> TypeError                                 Traceback (most recent call last)
> ```

As of now we assumed that the values given in the dictionary can be anyone of lists, functions or a single value. If we implement it to take in matrices also as values...yes we'll be able to get banded block matrices.

GYeyosi commented:
> As of now we assumed that the values given in the dictionary can be anyone of lists, functions or a single value. If we implement it to take in matrices also as values...yes we'll be able to get banded block matrices.

I tried implementing it, but it's very confusing on how the matrix produced should look like. As of now, I think it's better not adding this functionality.

smichr committed:
allow for {0} to make arrowhead matrices

- more error checking, especially to guard from overwriting values
- default to square matrix
- size can be an int: 'size': 1 ==> 'size': (1,1)

smichr committed:
remove {0}; recognize matrix entries

smichr commented:
```
    a = ones(2)
    assert diag({0:[1]*7, 3:a, -3:-a}) == Matrix([
        [ 1,  0,  0,  1, 1, 0, 0],
        [ 0,  1,  0,  1, 1, 0, 0],
        [ 0,  0,  1,  0, 0, 1, 1],
        [-1, -1,  0,  1, 0, 1, 1],
        [-1, -1,  0,  0, 1, 0, 0],
        [ 0,  0, -1, -1, 0, 1, 0],
        [ 0,  0, -1, -1, 0, 0, 1]])
```

pr pending

GYeyosi committed:
Merge pull request #4 from smichr/diag

smichr commented:
> reasonable way of using the features here for banded block matrices?
```python
>>> diag({-1:ones(2)}, ones(2))
Matrix([
[0, 0, 0, 0],
[1, 1, 0, 0],
[1, 1, 0, 0],
[0, 0, 1, 1],
[0, 0, 1, 1]])
```

smichr commented:
> it's very confusing

I agree. It will be worth testing this for a while to make sure there are no corners left unchecked.

smichr commented on sympy/matrices/common.py:
```diff
@@ -673,16 +674,19 @@ def diag(kls, *args, **kwargs):
-        >>> Matrix.diag([1, 2, 3])
+        >>> diag = Matrix.diag
+
+        All diagonal elements can be given individually
```
@siefkenj or @asmeurer, can you tell me why we have two routines that behave differently? From `sympy import *` state, if you use `diag` you get the version that comes from `dense` but if you use `Matrix.diag` you get this one. And this one unpacks a list whereas the former converts lists to Matrices. So in the former you would insert a column down from the diagonal while the in the latter you would get a diagonal of elements: 
```
[1, 0]     [1, 0]
[0. 1] vs [1, 0]
```

siefkenj replied:
I don't know why this is (other than it's the way it's been for a long time...). I think that they should have the same behavior, which is the one that `Matrix.diag` has. It's also weird that the docstring is repeated.

smichr replied:
During the renaming of mutable to dense and the cleanup and separation of matrices to common, the two version persisted and then during the cleanup you added unpacking to one and normalization to the other. So the public facing `diag` does not unpack and converts lists to matrices. I think that makes sense. The codebase also uses it this way when it calls `diag(*foo)`.
I won't break anything but I will get rid of the routine from `dense`.

smichr committed:
modify dense.diag docstring

siefkenj commented:
This PR looks really, really complex to me, and I don't understand the API. I don't think we should not attempt to support banded block-matrices (i.e., a matrix off the diagonal) and function arguments (though I could be convinced otherwise, it looks really complicated right now).

I would propose adding a `bands={...}` keyword-only argument and limit it to the form `{<band offset>: <number || list>}`. If someone wants to do complicated things with functions, use a function to create a matrix in the first place.

For example, 

```
def get_band(i,j):
    return j - i

Matrix(5,5, lambda i,j: get_band(i,j)**2)
```


As a separate issue, the code formatting of the unit tests should be a separate PR so this one can focus on just new code.

GYeyosi committed:
Merge pull request #5 from smichr/diag
